apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerReporter.java

```diff
@@ -166,7 +173,10 @@ public void scheduleWakeupEvent() {
     }
 
     @Override
-    public boolean flush(long timeout, TimeUnit unit) {
+    public boolean flush(long timeout, TimeUnit unit, boolean followupWithFlushRequest) {
+        if (followupWithFlushRequest) {
+            return publishAndWaitForEvent(timeout, unit, MAKE_FLUSH_REQUEST_EVENT_TRANSLATOR);
+        }
         return publishAndWaitForEvent(timeout, unit, END_REQUEST_EVENT_TRANSLATOR);
```

Comment:
There are a few potential issues when you have two events instead of one:
- The total blocked time may be higher: you may have to wait for the events to be flushed that were enqueued while waiting for the flush request event.
- In situations where the queue is almost full, one of the two events may make it into the queue.
- A related issue is increased wait time due to more re-tries when the queue is full. Essentially, the effective timeout is twice the requested timeout.

Therefore, I think it makes sense to just have one flushing event. We could even keep it simple and always both end the request and execute a flush request. As the only user for the flush event is the Lambda instrumentation, I don't think we need the `followupWithFlushRequest` argument. All production usages would provide `true`.

If a use case comes up where we do differ between end request and executing a flush request, we could add a boolean flag to the `ReportingEvent`. But for now, I'd keep it simple and always end the request, start a flush request and end the flush request.

Reply:
Not sure I understand what you mean, it's a single event - either end or end and flush.
The comment about continuing to the `END_REQUEST` case is only regarding the `switch`-`case` execution flow

Reply:
That's how I guaranteed atomicity

Reply:
But maybe we just need the flush request event and not the end request event. But I leave that up to you.

Reply:
Let's leave it as is for now